OpenSim/Simulation/SimbodyEngine/Joint.cpp

```diff
@@ -266,43 +266,56 @@ bool Joint::isCoordinateUsed(const Coordinate& aCoordinate) const
 
 void Joint::scale(const ScaleSet& scaleSet)
 {
+    // Scale the parent and/or child Frame, but only if owned by the Joint. The
```

Comment:
The previous implementation was overly complex/confusing and your fix can go a long way to simplify it. We can use `findComponent<PhysicalOffsetFrame>` to simplify further. I tested the following and it works:

```cpp
void Joint::scale(const ScaleSet& scaleSet)
{
    auto scaleOffsetFrameIfOwned = [&](const PhysicalFrame& frame) -> void {
        // If the frame is owned by this Joint then scale it.
        // Otherwise let the owner of the frame decide.
        if (auto* offset = findComponent<PhysicalOffsetFrame>(frame.getName()))
        {
            // find the scale factors associated with the base PhysicalFrame
            // of this offset frame
            const string& baseFrameName = offset->findBaseFrame().getName();

            // Get scale factors if they exist
            for (int i = 0; i < scaleSet.getSize(); i++) {
                Scale& scale = scaleSet.get(i);
                if (scale.getSegmentName() == baseFrameName) {
                    SimTK::Vec3 factors(1.0);
                    scale.getScaleFactors(factors);
                    const_cast<PhysicalOffsetFrame*>(offset)->scale(factors);
                    return;
                }
            }
        }
    };

    scaleOffsetFrameIfOwned(getParentFrame());
    scaleOffsetFrameIfOwned(getChildFrame());
}
```
Something like this might be easier to follow and it avoids the `dynamic_cast`.

Reply:
> The previous implementation was overly complex/confusing and your fix can go a long way to simplify it.

The purpose of this PR is to fix a bug and avoid searching through `scaleSet` unnecessarily.

> We can use `findComponent<PhysicalOffsetFrame>` to simplify further. I tested the following and it works:

I believe that `findComponent()` would work but I disagree with using it here. I think the intent of `dynamic_cast` is different from the intent of `findComponent()` (it doesn't matter that they produce the same result). We already have references to the parent and child Frames; there is no need to find them. What we need to do is check whether these Frames are PhysicalOffsetFrames, and it appears that the convention in OpenSim is to use `dynamic_cast` to do this. A reasonable implementation of `findComponent()` could be to navigate to the provided absolute path and check whether the Component exists and is of type PhysicalOffsetFrame; if not, continue looking around for a Component of type PhysicalOffsetFrame of the provided name (e.g., in the ownership subtree below the provided absolute path). I believe that is the semantics of a "find" operation and would be the interpretation of such a line of code, but we don't want to do that here lest we accidentally find some Frame other than the parent/child.

> Something like this might be easier to follow and it avoids the `dynamic_cast`.

I think `findComponent()` is more difficult to follow because you are searching for something you already have. Please explain why `dynamic_cast` is acceptable elsewhere in the repo but not here. It would help me to understand your objective function. Are you trying to avoid using `dynamic_cast` for some reason? Are you trying to minimize the number of lines of code?

Reply:
> and it avoids the `dynamic_cast`

Also unclear why `dynamic_cast` should be avoided here since `findComponent()` uses `dynamic_cast` anyway. If you don't want `dynamic_cast` in this method, we could move that piece of `findComponent()` (i.e., `found = dynamic_cast<const C*>(this); if (found) return found;`) to a new method like `getAs<PhysicalOffsetFrame*>(absolute_path)` and call `getAs` on line 274.

Reply:
> The purpose of this PR is to fix a bug and improve performance (by avoiding unnecessary searches through scaleSet). I agree with simplifying code in general, but I don't think the Development Team can afford the time required to make that a priority.

I provided an alternative that maintains the purpose of the PR and IMHO is also more readable. I this case both are achievable.

>I believe that findComponent() would work but I disagree with using it here. I think the intent of dynamic_cast is different from the intent of findComponent() (it doesn't matter that they currently produce the same result) ...  but that's the wrong idea: we don't want to do that here lest we accidentally find some Frame other than the parent/child.

Your current code will not find the parent/child `PhysicalOffsetFrame` that was added by `addComponent` to the `Joint`. That was part of the criticism of the previous implementation. You are trying to *find* the parent/child `PhysicalOffsetFrame` that is owned by the `Joint`, which IMHO is the purview of `findComponent`.  You could thwart the issue of finding some other `PhysicalOffsetFrame` in the `Joint`  with the same name as the child/parent frame, by verifying its immediate `owner` is the `Joint`. I agree your current use of property's `findIndex()` ensure it is an immediate subcomponent of the `Joint` but it does not allow the child/parent to have been add via `addComponent`. 

> Also unclear why dynamic_cast should be avoided here since findComponent() uses dynamic_cast anyway

That's the point. Why pollute the implementation with your own find and dynamic cast  when there is a utility method `findComponent` to do it for you.  

To me these implementation details are not worth this level of argument.  I made a suggestion to be helpful. If you strongly believe there is nothing helpful in suggestion if the other reviewer is fine with your implementation I am too. Let's move on.

Reply:
> I provided an alternative that maintains the purpose of the PR and IMHO is also more readable. I this case both are achievable.

I find `dynamic_cast` more readable because it's immediately obvious what it does; `findComponent()` could do all sorts of things (including tree traversals, incurring overhead, etc.) that I think would be best to avoid if possible.

> Your current code will not find the parent/child `PhysicalOffsetFrame` that was added by `addComponent` to the `Joint`. That was part of the criticism of the previous implementation.

The criticism of the previous implementation is stated in #1913: although it is not currently covered by a test case, `Joint::scale()` will throw an exception (or have unexpected behavior) in certain situations. This PR is not intended (or advertised) to address the larger design flaw reported in #1923. Fixing that issue is a big job and will require design work, test case development, etc. Currently, the car doesn't work; I don't think it makes sense to worry about the fact that the windshield wipers don't work since (in this analogy) they will probably need to be removed to fix the car—and the final design might not need windshield wipers. This PR is mainly trying to keep the wheels from falling off.

> You are trying to *find* the parent/child `PhysicalOffsetFrame` that is owned by the `Joint`, which IMHO is the purview of `findComponent`.  You could thwart the issue of finding some other `PhysicalOffsetFrame` in the `Joint`  with the same name as the child/parent frame, by verifying its immediate `owner` is the `Joint`. I agree your current use of property's `findIndex()` ensure it is an immediate subcomponent of the `Joint` but it does not allow the child/parent to have been add via `addComponent`.

Using `findComponent()` and confirming that the immediate owner is the Joint would indeed address the situation you mention, but what if the Joint owns a component (added via `addComponent()`) that then owns a PhysicalOffsetFrame to which the Joint connects? I think the scaling infrastructure is more than a couple tweaks away from working and requires a more rigorous approach.

Reply:
> I think the scaling infrastructure is more than a couple tweaks away from working and requires a more rigorous approach.

I totally agree. It would be great to work with you on it!